### PR TITLE
Feature/acurate tracking

### PR DESCRIPTION
**Pull request by jbout**
Ensure we capture queries triggered by QueryBuilder::execute()

### CODE REVIEW

File: models/persistence/QueryCounter.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    private $functions = [];
+
+    public function __construct(string $persistenceId) {
+        $this->id = $persistenceId;
+    }
+
+    function count(string $functionName, string $sqlStatement): void
+    {
+        //$this->logDebug('    '.$functionName.' '.substr(str_replace('\n', ' ', $sqlStatement), 0, 100));
```
Comment (mike-ionut-mihai-sandu-tao) [inline models/persistence/QueryCounter.php:43]: - [ ] let's remove the comment

File: models/persistence/QueryBuilderProxy.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    /** @var QueryCounter */
+    private $counter;
+
+    public function __construct(Connection $connection, QueryCounter $counter)
+    {
+        parent::__construct($connection);
+        $this->counter = $counter;
+    }
+
+    public function execute()
```
Comment (mike-ionut-mihai-sandu-tao) [inline models/persistence/QueryBuilderProxy.php:38]: return type

File: models/persistence/QueryBuilderProxy.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+use Doctrine\DBAL\Connection;
+
+class QueryBuilderProxy extends QueryBuilder
+{
+    /** @var QueryCounter */
+    private $counter;
+
+    public function __construct(Connection $connection, QueryCounter $counter)
+    {
+        parent::__construct($connection);
```
Comment (mike-ionut-mihai-sandu-tao) [inline models/persistence/QueryBuilderProxy.php:34]: type hint

File: models/persistence/PlatformProxy.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2017 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```
Comment (mike-ionut-mihai-sandu-tao) [inline models/persistence/PlatformProxy.php:18]: 2020

File: models/persistence/PlatformProxy.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    /** @var QueryCounter */
+    private $counter;
+
+    public function __construct($dbalConnection, QueryCounter $counter)
+    {
+        parent::__construct($dbalConnection);
+        $this->counter = $counter;
+    }
+
+    public function getQueryBuilder()
```
Comment (mike-ionut-mihai-sandu-tao) [inline models/persistence/PlatformProxy.php:35]: return type

File: models/persistence/QueryBuilderProxy.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2017 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
```
Comment (mike-ionut-mihai-sandu-tao) [inline models/persistence/QueryBuilderProxy.php:18]: 2020 year

File: models/persistence/QueryBuilderProxy.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ */
+
+namespace oat\taoDevTools\models\persistence;
+
+use Doctrine\DBAL\Query\QueryBuilder;
```
Comment (mike-ionut-mihai-sandu-tao) [inline models/persistence/QueryBuilderProxy.php:24]: should we add Doctrine as a dependency inside this extension ?

File: models/persistence/QueryCounter.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        $this->count++;
+        $this->functions[$functionName] = isset($this->functions[$functionName])
+            ? $this->functions[$functionName] + 1
+            : 1
+        ;
+    }
+
+    public function __destruct()
```
Comment (mike-ionut-mihai-sandu-tao) [inline models/persistence/QueryCounter.php:51]: are you sure you need this ?

Comment (jbout) [reply models/persistence/QueryCounter.php:52]: this part does the actual logging

File: models/persistence/QueryBuilderProxy.php (hunk 3 above)
Comment (jbout) [reply models/persistence/QueryBuilderProxy.php:34]: added doc block instead, strong typing not possible

File: models/persistence/QueryCounter.php (hunk 1 above)
Comment (jbout) [reply models/persistence/QueryCounter.php:44]: would like to leave it for easy debugging, could comment the comment instead?

File: models/persistence/QueryBuilderProxy.php (hunk 7 above)
Comment (jbout) [reply models/persistence/QueryBuilderProxy.php:24]: I would not add it to devtools as well as this would make it harder to upgrade minor versions

**Pull request merged**